src/backend/distributed/executor/executor.c

```diff
@@ -888,6 +898,17 @@ AssignTasksToConnections(DistributedExecution *execution)
 					/* task is ready to execute on any session */
 					dlist_push_tail(&workerPool->readyTaskQueue,
 									&placementExecution->workerReadyQueueNode);
+
+					/*
+					 * Record the first task assignment time to the pool. We need this
+					 * to enforce NodeConnectionTimeout.
+					 */
+					if (workerPool->activeConnectionCount == 0 &&
+						workerPool->readyTaskCount == 0)
```

Comment:
I'm a bit confused why we care about the task assignment. Why not first connection establishment?

Reply:
> Why not first connection establishment?

Do you mean the pool creation as I used to do?

When replication factor > 1 and single task query, we create the pools at the same time but only try the first pool to access the placement until the timeout. So, the time that actually matters seems when the first task becomes ready to execute on the pool.

Reply:
but we don't open a connection if there are no ready tasks, right?

I was expecting it to be somewhere around here:
https://github.com/citusdata/citus/blob/32678f8f871e2324dcb2446010946d92457a81aa/src/backend/distributed/executor/executor.c#L1298

I'm also wondering whether it's worth marking sessions as "critical" to make the distinction between critical and non-critical connection establishment a bit more obvious to the reader.

Reply:
> I was expecting it to be somewhere around here:

Ok, that makes sense. I agree that it is not easier to follow/understand the code.

> I'm also wondering whether it's worth marking sessions as "critical" to make the distinction between critical and non-critical connection establishment a bit more obvious to the reader.

Do you mean on #2713 get rid of (or rely less on) `errorOnAnyFailure ` and instead mark connections critical? Let me also think about that a bit more if that's what you're asking.